# Обучение моделей P1P2

##### Загрузка данных

In [18]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 50)
pd.set_option('display.precision', 4)

RANDOM_STATE = 1337

In [19]:
from acquiring_xsell.data import load_data
from acquiring_xsell.preprocessing import process_dtypes, fill_missing_values
from acquiring_xsell.features import build_features, make_product_df

df = load_data()
df = process_dtypes(df)
df = fill_missing_values(df)
df = build_features(df)
df = make_product_df(df, "p1p2")

df.head()

,month,inn_status,current_segment_inn,top_mcc_group_inn,who_is_this_first_inn,lifetime_month_streak_inn,real_kam_on_inn,has_termination_start,altpay1_turnover,altpay2_turnover,altpay3_turnover,altpay4_turnover,altpay5_turnover,altpay6_turnover,has_recurrent_payments,cnt_trx,turnover,other_active_product_cnt,is_relevant_mcc_p3,is_relevant_mcc_altpay5,is_relevant_turnover_altpay5,active_mcc_cnt,active_gateway_cnt,avg_check,turnover_ma_2m,turnover_ema_2m,turnover_ma_3m,turnover_ema_3m,turnover_ma_6m,turnover_ema_6m,cnt_trx_ma_2m,cnt_trx_ema_2m,cnt_trx_ma_3m,cnt_trx_ema_3m,cnt_trx_ma_6m,cnt_trx_ema_6m,avg_check_wma_2m,avg_check_wma_3m,avg_check_wma_6m,turnover_cv_2m,turnover_cv_3m,turnover_cv_6m,turnover_change_to_prev_month,cnt_trx_change_to_prev_month,avg_check_change_to_prev_month,turnover_change_to_turnover_ma_2m,turnover_change_to_turnover_ma_3m,turnover_change_to_turnover_ma_6m,turnover_change_to_turnover_ema_2m,turnover_change_to_turnover_ema_3m,turnover_change_to_turnover_ema_6m,cnt_trx_change_to_cnt_trx_ma_2m,cnt_trx_change_to_cnt_trx_ma_3m,cnt_trx_change_to_cnt_trx_ma_6m,cnt_trx_change_to_cnt_trx_ema_2m,cnt_trx_change_to_cnt_trx_ema_3m,cnt_trx_change_to_cnt_trx_ema_6m,avg_check_change_to_avg_check_wma_2m,avg_check_change_to_avg_check_wma_3m,avg_check_change_to_avg_check_wma_6m,is_first_trx_month,is_first_history_month,months_since_last_active_month,months_since_first_trx,has_p3,has_altpay5,rule_flag_p1p2,rule_flag_p3,rule_flag_altpay5,became_active_recently,target
0,2025-04-01,Churn,XXS,Tickets&clubs,referral,0,1,0,0.0,0.0,0.0000,0.0,0.0,0.0,0,0.0000,0.0000,0.0,0,0,0,0,0,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0,0,9,48,0,0,0,0,0,0,0
1,2025-04-01,Churn,XXS,Travel,outcoming,0,1,0,0.0,0.0,0.0000,0.0,0.0,0.0,0,0.0000,0.0000,0.0,0,0,0,0,0,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0,0,8,33,0,0,0,0,0,0,0
2,2025-04-01,Active,XXS,Digital,outcoming,33,0,0,0.0,0.0,0.0000,0.0,0.0,0.0,0,0.0063,0.0009,0.0,0,1,0,1,1,0.0005,0.0044,0.0038,0.0024,0.0035,0.0029,0.0019,0.0103,0.0052,0.0110,0.0144,0.0073,0.0079,0.0016,0.0009,0.0008,0.7867,0.9640,0.6125,-0.7143,-0.3750,-0.5429,-0.5556,-0.4000,-0.5000,-0.3632,-0.4206,-0.4641,-0.2308,0.0000,-0.0385,-0.0833,-0.0774,-0.0577,-0.4222,-0.4000,-0.4800,0,0,0,32,0,0,1,0,0,0,0
3,2025-04-01,Active,M,Digital,outcoming,17,1,0,0.0,0.0,0.0000,0.0,0.0,0.0,0,0.6378,0.6075,0.0,0,1,1,1,1,0.0418,0.6193,0.6119,0.6091,0.6117,0.6369,0.6182,0.6069,0.6321,0.6532,0.6372,0.6257,0.6423,0.0628,0.0382,0.0446,0.0120,0.0634,0.1056,0.0172,0.0270,-0.0096,0.0085,0.0450,-0.0202,0.0072,0.0055,-0.0035,0.0133,0.0703,0.1063,0.0225,0.0418,0.0700,-0.0047,-0.0236,-0.1144,0,0,0,16,0,0,1,0,1,0,0
4,2025-04-01,Active,XS,Charity,referral,36,0,0,0.0,0.0,0.4429,0.0,0.0,0.0,1,0.1768,0.4405,0.0,1,0,1,2,2,0.2072,0.4681,0.4622,0.4383,0.4652,0.4628,0.4501,0.1801,0.1539,0.2115,0.2397,0.1922,0.2072,0.2443,0.1825,0.1970,0.1541,0.1917,0.2269,0.2447,0.0302,0.2081,0.1090,0.2055,0.2004,0.0742,0.1112,0.1537,0.0149,0.0342,0.0119,0.0097,0.0119,0.0124,0.0927,0.1656,0.1863,0,0,0,35,0,0,0,0,0,0,0


## Метрики качества
Основной метрикой качества моделей является Precision@K, поскольку бизнес-задача заключается в эффективном распределении ограниченного ресурса отдела продаж. В рамках задачи capacity отдела составляет 2000 предложений в месяц, что определяет основное значение K. Дополнительно рассматриваются значения K=500, K=1000 и K=1500 для анализа устойчивости качества ранжирования.

Для оценки относительного качества моделей используется метрика Uplift — отношение Precision@K рассматриваемой модели к Precision@K базовых подходов. В качестве базовых подходов используются:

1. Случайный отбор клиентов
2. Отбор на основе текущих бизнес-правил

Дополнительно для оценки качества ранжирования анализируется зависимость Precision@K от K, что позволяет оценить способность модели выделять наиболее релевантных клиентов в верхней части ранжированного списка.

Также используется метрика Recall@K, отражающая долю всех целевых событий (подключений продукта), попадающих в top-K клиентов, что позволяет оценить полноту охвата моделью потенциально релевантных клиентов.

Для интерпретации результатов в бизнес-терминах рассчитывается ожидаемое количество подключений (Expected Conversions), определяемое как произведение Precision@K на K.

## Случайный отбор


Для случайного отбора ожидаемое значение Precision@K совпадает с долей положительных объектов в выборке. В связи с этим в качестве baseline используется доля положительного класса, рассчитанная по каждому месяцу.

In [3]:
from acquiring_xsell.preprocessing import train_test_month_split
from acquiring_xsell.metrics import calculate_random_baseline

train, test = train_test_month_split(df)

print("Random Selection P1P2")

print("\nTrain")
calculate_random_baseline(train)

print("\nTest")
calculate_random_baseline(test)

Random Selection P1P2

Train
K=500  | Expected Precision: 0.22% | Selection Rate: 2.24% | Expected Conversions: 1.12
K=1000 | Expected Precision: 0.22% | Selection Rate: 4.48% | Expected Conversions: 2.25
K=1500 | Expected Precision: 0.22% | Selection Rate: 6.72% | Expected Conversions: 3.37
K=2000 | Expected Precision: 0.22% | Selection Rate: 8.97% | Expected Conversions: 4.50

Test
K=500  | Expected Precision: 0.25% | Selection Rate: 2.17% | Expected Conversions: 1.24
K=1000 | Expected Precision: 0.25% | Selection Rate: 4.34% | Expected Conversions: 2.49
K=1500 | Expected Precision: 0.25% | Selection Rate: 6.50% | Expected Conversions: 3.73
K=2000 | Expected Precision: 0.25% | Selection Rate: 8.67% | Expected Conversions: 4.97


## Business-Rule

Критерии релевантности:
- Отсутствие подключённых продуктов `P1`, `P2` (фильтрация уже реализована в `df_p1_p2`)
- Статус `'Active'`, `'Reborn'`
- Группа MCC `!= 'Charity'`

Кандидаты на подключение ранжируется по убыванию `turnover`

Флаг уже есть в датасете. Замерим качество.

In [4]:
from acquiring_xsell.preprocessing import train_test_month_split
from acquiring_xsell.metrics import calculate_metrics

train, test = train_test_month_split(df)

y_score = 'turnover'

print(f"Business-Rule Selection P1P2, y_score={y_score}")

print("\nTrain")
calculate_metrics(data=train, y_score=y_score, rule_flag_col='rule_flag_p1p2')

print("\nTest")
calculate_metrics(data=test, y_score=y_score, rule_flag_col='rule_flag_p1p2')

Business-Rule Selection P1P2, y_score=turnover

Train
K=500  | Precision: 0.67% | Recall: 6.91% | Expected Conversions: 3.33
K=1000 | Precision: 0.63% | Recall: 13.11% | Expected Conversions: 6.25
K=1500 | Precision: 0.57% | Recall: 18.15% | Expected Conversions: 8.58
K=2000 | Precision: 0.52% | Recall: 21.16% | Expected Conversions: 10.33

Test
K=500  | Precision: 0.27% | Recall: 2.21% | Expected Conversions: 1.33
K=1000 | Precision: 0.30% | Recall: 5.36% | Expected Conversions: 3.00
K=1500 | Precision: 0.36% | Recall: 9.55% | Expected Conversions: 5.33
K=2000 | Precision: 0.43% | Recall: 15.32% | Expected Conversions: 8.67


Попробуем улучшить качество отбора по бизнес-правилу за счёт альтернативной метрики ранжирования. Переберём все числовые признаки, кроме таргета. Также попробуем ранжировать не только по убыванию скора, но и по возрастанию.

In [5]:
from acquiring_xsell.preprocessing import train_test_month_split
from acquiring_xsell.metrics import calculate_metrics

train, test = train_test_month_split(df)

num_cols = train.select_dtypes(include='number').columns

print("Business-Rule Selection P1P2, scoring by all numeric features")

for col in num_cols:
    if col != 'target':
        for ascending in [True, False]:
            print("\n", "="*10, f"y_score='{col}', ascending={ascending}", "="*10)

            print(f"\nTrain")
            calculate_metrics(data=train, y_score=col, rule_flag_col='rule_flag_p1p2', ascending=ascending)

            print(f"\nTest")
            calculate_metrics(data=test, y_score=col, rule_flag_col='rule_flag_p1p2', ascending=ascending)

Business-Rule Selection P1P2, scoring by all numeric features

 ========== y_score='lifetime_month_streak_inn', ascending=True ==========

Train
K=500  | Precision: 2.72% | Recall: 25.57% | Expected Conversions: 13.58
K=1000 | Precision: 1.76% | Recall: 33.66% | Expected Conversions: 17.58
K=1500 | Precision: 1.32% | Recall: 38.31% | Expected Conversions: 19.75
K=2000 | Precision: 1.07% | Recall: 41.62% | Expected Conversions: 21.33

Test
K=500  | Precision: 3.13% | Recall: 27.33% | Expected Conversions: 15.67
K=1000 | Precision: 2.20% | Recall: 38.11% | Expected Conversions: 22.00
K=1500 | Precision: 1.71% | Recall: 44.63% | Expected Conversions: 25.67
K=2000 | Precision: 1.40% | Recall: 48.68% | Expected Conversions: 28.00

 ========== y_score='lifetime_month_streak_inn', ascending=False ==========

Train
K=500  | Precision: 0.25% | Recall: 2.49% | Expected Conversions: 1.25
K=1000 | Precision: 0.34% | Recall: 6.70% | Expected Conversions: 3.42
K=1500 | Precision: 0.29% | Recall: 8.5

Лучшие метрики на тесте при K=2000 выявлены следующих скоринг-признаков:
```
y_score='months_since_first_trx', ascending=True
Precision: 1.43% | Recall: 49.60% | Expected Conversions: 28.67

y_score='lifetime_month_streak_inn', ascending=True
Precision: 1.40% | Recall: 48.68% | Expected Conversions: 28.00

y_score='became_active_recently', ascending=False
Precision: 1.33% | Recall: 46.47% | Expected Conversions: 26.67

y_score='is_relevant_turnover_altpay5', ascending=True 
Precision: 1.33% | Recall: 46.44% | Expected Conversions: 26.67

y_score='real_kam_on_inn', ascending=True
Precision: 1.32% | Recall: 45.41% | Expected Conversions: 26.33

y_score='is_relevant_mcc_altpay5', ascending=False
Precision: 1.30% | Recall: 45.07% | Expected Conversions: 26.00
```
Рассмотрим их отдельно

In [6]:
from acquiring_xsell.preprocessing import train_test_month_split
from acquiring_xsell.metrics import calculate_metrics

train, test = train_test_month_split(df)

best_business_rule_scoring_features = {
    'months_since_first_trx': True,
    'lifetime_month_streak_inn': True,
    'became_active_recently': False,
    'is_relevant_turnover_altpay5': True,
    'real_kam_on_inn': True,
    'is_relevant_mcc_altpay5': False,
}

print("Business-Rule Selection P1P2")

for col, ascending in best_business_rule_scoring_features.items():
    print("\n", "="*10, f"y_score='{col}', ascending={ascending}", "="*10)
    
    print(f"\nTrain")
    calculate_metrics(data=train, y_score=col, rule_flag_col='rule_flag_p1p2', ascending=ascending)
    
    print(f"\nTest")
    calculate_metrics(data=test, y_score=col, rule_flag_col='rule_flag_p1p2', ascending=ascending)

Business-Rule Selection P1P2

 ========== y_score='months_since_first_trx', ascending=True ==========

Train
K=500  | Precision: 2.88% | Recall: 27.73% | Expected Conversions: 14.42
K=1000 | Precision: 1.68% | Recall: 32.65% | Expected Conversions: 16.83
K=1500 | Precision: 1.19% | Recall: 34.79% | Expected Conversions: 17.92
K=2000 | Precision: 1.02% | Recall: 39.84% | Expected Conversions: 20.33

Test
K=500  | Precision: 3.80% | Recall: 32.84% | Expected Conversions: 19.00
K=1000 | Precision: 2.37% | Recall: 41.00% | Expected Conversions: 23.67
K=1500 | Precision: 1.78% | Recall: 46.18% | Expected Conversions: 26.67
K=2000 | Precision: 1.43% | Recall: 49.60% | Expected Conversions: 28.67

 ========== y_score='lifetime_month_streak_inn', ascending=True ==========

Train
K=500  | Precision: 2.72% | Recall: 25.57% | Expected Conversions: 13.58
K=1000 | Precision: 1.76% | Recall: 33.66% | Expected Conversions: 17.58
K=1500 | Precision: 1.32% | Recall: 38.31% | Expected Conversions: 19.75

Лучший по качеству признак для ранжирования — `months_since_first_trx` при сортировке по возрастанию. Он лучше всего ранжирует верхнюю часть списка (Precision@500 = 3.80%) и даёт наибольшее кол-во ожидаемых конверсий при всех K в сравнении с остальными. Это подтверждает гипотезу из EDA о том, что клиенты, которые недавно совершили первую транзакцию по основному продукту, наиболее предрасположены к подключению P1P2. Также схожие значения метрик у скоринг-признаков `lifetime_month_streak_inn` и `became_active_recently` подтверждают эту гипотезу, т.к. указывают на схожие клиентские характеристики.

## Logistic Regression

Обучим логистическую регрессию на всех признаках. В рамках Pipeline применим One-Hot Encoding для категориальных признаков, масштабирование числовых признаков через StandardScaler и L1-регуляризацию, чтобы модель с её помощью могла самостоятельно сделать feature selection.

In [7]:
from acquiring_xsell.preprocessing import train_test_month_split
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from acquiring_xsell.metrics import calculate_metrics


train, test = train_test_month_split(df)

X_train = train.drop(columns=['target', 'month'])
y_train = train['target']

X_test = test.drop(columns=['target', 'month'])
y_test = test['target']


num_cols = X_train.select_dtypes(include='number').columns
cat_cols = X_train.select_dtypes(include=['object', 'category']).columns

column_transformer = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), num_cols),
        ('cat', OneHotEncoder(drop='first', sparse_output=True, dtype=np.int8), cat_cols),
    ],
    remainder='passthrough'
)


c_values = [0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 5, 10]
lr_models = {}

print(f"Logistic Regression P1P2")

for c in c_values:
    lr_pipe = Pipeline([
        ('column_transformer', column_transformer),
        ('lr', LogisticRegression(
                        max_iter=100_000,
                        solver="liblinear",
                        l1_ratio=1,
                        C=c,
                        random_state=RANDOM_STATE
                    ))
    ])

    lr_pipe.fit(X_train, y_train)

    lr_models[c] = lr_pipe

    train_eval = train.copy()
    train_eval['lr_proba'] = lr_pipe.predict_proba(X_train)[:, 1]
    
    test_eval = test.copy()
    test_eval['lr_proba'] = lr_pipe.predict_proba(X_test)[:, 1]

    print("\n", "="*30, f"C={c}", "="*30)

    print(f"\nTrain")
    calculate_metrics(data=train_eval, y_score='lr_proba')

    print(f"\nTest")
    calculate_metrics(data=test_eval, y_score='lr_proba')

Logistic Regression P1P2

 ============================== C=0.001 ==============================

Train
K=500  | Precision: 2.17% | Recall: 20.41% | Expected Conversions: 10.83
K=1000 | Precision: 1.16% | Recall: 21.79% | Expected Conversions: 11.58
K=1500 | Precision: 0.84% | Recall: 23.65% | Expected Conversions: 12.58
K=2000 | Precision: 0.70% | Recall: 26.69% | Expected Conversions: 14.08

Test
K=500  | Precision: 2.93% | Recall: 25.33% | Expected Conversions: 14.67
K=1000 | Precision: 1.63% | Recall: 28.17% | Expected Conversions: 16.33
K=1500 | Precision: 1.20% | Recall: 31.14% | Expected Conversions: 18.00
K=2000 | Precision: 0.93% | Recall: 32.24% | Expected Conversions: 18.67

 ============================== C=0.005 ==============================

Train
K=500  | Precision: 2.88% | Recall: 27.26% | Expected Conversions: 14.42
K=1000 | Precision: 1.76% | Recall: 33.16% | Expected Conversions: 17.58
K=1500 | Precision: 1.29% | Recall: 37.12% | Expected Conversions: 19.33
K=2000 |

Лучшее качество логистической регрессии достигнуто при гиперпараметре C=0.05. По Precision@2000 это лучше случайного отбора в 6 раз и в 3.5 раза лучше бейзлайна отбора по бизнес-правилу. Это также чуть лучше отбора по бизнес-правилу при скоринге по признаку `months_since_first_trx` (LogReg 1.52 vs. 1.43 BR).

```
Train
K=500  | Precision: 3.10% | Recall: 29.72% | Expected Conversions: 15.50
K=1000 | Precision: 1.88% | Recall: 36.21% | Expected Conversions: 18.75
K=1500 | Precision: 1.39% | Recall: 40.63% | Expected Conversions: 20.92
K=2000 | Precision: 1.18% | Recall: 46.44% | Expected Conversions: 23.58

Test
K=500  | Precision: 3.80% | Recall: 33.02% | Expected Conversions: 19.00
K=1000 | Precision: 2.37% | Recall: 40.88% | Expected Conversions: 23.67
K=1500 | Precision: 1.73% | Recall: 45.19% | Expected Conversions: 26.00
K=2000 | Precision: 1.52% | Recall: 52.87% | Expected Conversions: 30.33
```

Посмотрим на веса модели.

In [8]:
def print_lr_coefficients(lr_model):
    feature_names = []

    num_cols = lr_model.named_steps['column_transformer'].transformers_[0][2]
    cat_encoder = lr_model.named_steps['column_transformer'].transformers_[1][1]
    cat_cols = lr_model.named_steps['column_transformer'].transformers_[1][2]

    feature_names.extend(num_cols)
    cat_feature_names = cat_encoder.get_feature_names_out(cat_cols)
    feature_names.extend(cat_feature_names)

    coefs = lr_model.named_steps['lr'].coef_[0]
    coef_df = pd.DataFrame({'feature': feature_names, 'coef': coefs})

    top_10 = coef_df.sort_values('coef', ascending=False).head(10)
    bottom_10 = coef_df.sort_values('coef', ascending=True).head(10)
    zero_coefs = coef_df[coef_df['coef'] == 0]

    print("Топ-10 признаков с наибольшими coef_:")
    print(top_10.to_string(index=False))

    print("\nТоп-10 признаков с наименьшими coef_:")
    print(bottom_10.to_string(index=False))

    zero_coef_share = len(zero_coefs) / len(coef_df)
    print(f"\nДоля признаков с коэффициентом 0: {zero_coef_share:.2%}")

    print("\nПризнаки с коэффициентом 0:")
    if zero_coefs.empty:
        print("Нет признаков с коэффициентом 0.")
    else:
        print(zero_coefs.to_string(index=False))


print_lr_coefficients(lr_models[0.05])

Топ-10 признаков с наибольшими coef_:
                     feature   coef
              rule_flag_p1p2 0.2380
       top_mcc_group_inn_EDU 0.2123
          is_first_trx_month 0.1848
     is_relevant_mcc_altpay5 0.1766
              turnover_cv_6m 0.1192
is_relevant_turnover_altpay5 0.1018
                     cnt_trx 0.0861
      became_active_recently 0.0808
            altpay3_turnover 0.0521
            altpay1_turnover 0.0353

Топ-10 признаков с наименьшими coef_:
                        feature    coef
          has_termination_start -0.6554
     top_mcc_group_inn_FoodTech -0.4250
        top_mcc_group_inn_Churn -0.2615
      top_mcc_group_inn_Charity -0.2358
who_is_this_first_inn_outcoming -0.1648
                      avg_check -0.1420
               inn_status_Churn -0.1404
                turnover_ema_3m -0.1288
      lifetime_month_streak_inn -0.0902
         is_first_history_month -0.0659

Доля признаков с коэффициентом 0: 66.33%

Признаки с коэффициентом 0:
                

Модель занулила 66% признаков при обучении с L1-регуляризацией. Среди самых важных выделены:
- `has_termination_start = -0.6554` — выглядит логично, если компания запустила процесс ликвидации (т.е. закрывается), вряд ли она будет подключать дополнительные продукты.
- `top_mcc_group_inn = EDU` выделена как наиболее релевантная для продукта P1P2, в то время как `FoodTech` и `Charity` — наиболее нерелевантные. Наличие в отрицательном топе `top_mcc_group_inn` значения `Churn` (и `inn_status_Churn`) также объяснимо, если ИНН в статусе отток, вероятность того что она подключит P1P2 ниже, это подтверждает EDA.
- Также сильные веса имеют признаки, связанные с новизной клиента: `is_first_trx_month`, `became_active_recently`, `lifetime_month_streak_inn`. Первые 2 имеют положительные веса, последний — отрицательный.
- большой вес признака `rule_flag_p1p2` показывает, что по мнению модели существующее бизнес-правило имеет значительный положительный вес в определении положительного класса.
- Также интересными выглядят взаимосвязи с другими altpay-методами, большая часть из них была занулена, однако altpay1 и altpay3 находятся в топ-10 положительных весов.





В процессе анонимизации признаки `['turnover', 'cnt_trx', 'avg_check']` и их производные были масштабированы с использованием глобального `MinMaxScaler`. Это приводит к снижению сопоставимости значений во времени, к тому же практически все из них были убраны в ноль L-1 регуляризацией, поэтому также для сравнения обучим логистическую регрессию без данных признаков.

In [9]:
from acquiring_xsell.preprocessing import train_test_month_split
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from acquiring_xsell.metrics import calculate_metrics

PRE_SCALED_FEATURES = [
    'cnt_trx',
    'turnover',
    'avg_check',
    'turnover_ma_2m',
    'turnover_ema_2m',
    'turnover_ma_3m',
    'turnover_ema_3m',
    'turnover_ma_6m',
    'turnover_ema_6m',
    'cnt_trx_ma_2m',
    'cnt_trx_ema_2m',
    'cnt_trx_ma_3m',
    'cnt_trx_ema_3m',
    'cnt_trx_ma_6m',
    'cnt_trx_ema_6m',
    'avg_check_wma_2m',
    'avg_check_wma_3m',
    'avg_check_wma_6m',
    'altpay1_turnover',
    'altpay2_turnover',
    'altpay3_turnover',
    'altpay4_turnover',
    'altpay5_turnover',
    'altpay6_turnover',
]


train, test = train_test_month_split(df)

X_train = train.drop(columns=['target', 'month'] + PRE_SCALED_FEATURES)
y_train = train['target']

X_test = test.drop(columns=['target', 'month'] + PRE_SCALED_FEATURES)
y_test = test['target']


num_cols = X_train.select_dtypes(include='number').columns
cat_cols = X_train.select_dtypes(include=['object', 'category']).columns

column_transformer = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), num_cols),
        ('cat', OneHotEncoder(drop='first', sparse_output=True, dtype=np.int8), cat_cols),
    ],
    remainder='passthrough'
)


c_values = [0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 5, 10]
lr_models_wo_scaled_featuers = {}

print(f"Logistic Regression P1P2 without scaled features, y_score='lr_proba'")

for c in c_values:
    lr_pipe = Pipeline([
        ('column_transformer', column_transformer),
        ('lr', LogisticRegression(
                        max_iter=100_000,
                        solver="liblinear",
                        l1_ratio=1,
                        C=c,
                        random_state=RANDOM_STATE
                    ))
    ])

    lr_pipe.fit(X_train, y_train)

    lr_models_wo_scaled_featuers[c] = lr_pipe

    train_eval = train.copy()
    train_eval['lr_proba'] = lr_pipe.predict_proba(X_train)[:, 1]
    
    test_eval = test.copy()
    test_eval['lr_proba'] = lr_pipe.predict_proba(X_test)[:, 1]

    print("\n", "="*30, f"C={c}", "="*30)

    print(f"\nTrain")
    calculate_metrics(data=train_eval, y_score='lr_proba')

    print(f"\nTest")
    calculate_metrics(data=test_eval, y_score='lr_proba')

Logistic Regression P1P2 without scaled features, y_score='lr_proba'

 ============================== C=0.001 ==============================

Train
K=500  | Precision: 2.17% | Recall: 20.41% | Expected Conversions: 10.83
K=1000 | Precision: 1.16% | Recall: 21.79% | Expected Conversions: 11.58
K=1500 | Precision: 0.84% | Recall: 23.65% | Expected Conversions: 12.58
K=2000 | Precision: 0.70% | Recall: 26.69% | Expected Conversions: 14.08

Test
K=500  | Precision: 2.93% | Recall: 25.33% | Expected Conversions: 14.67
K=1000 | Precision: 1.63% | Recall: 28.17% | Expected Conversions: 16.33
K=1500 | Precision: 1.20% | Recall: 31.14% | Expected Conversions: 18.00
K=2000 | Precision: 0.93% | Recall: 32.24% | Expected Conversions: 18.67

 ============================== C=0.005 ==============================

Train
K=500  | Precision: 2.88% | Recall: 27.26% | Expected Conversions: 14.42
K=1000 | Precision: 1.76% | Recall: 33.16% | Expected Conversions: 17.58
K=1500 | Precision: 1.29% | Recall: 3

Лучшее качество достигается при C=0.1:

```
Train
K=500  | Precision: 3.02% | Recall: 28.80% | Expected Conversions: 15.08
K=1000 | Precision: 1.88% | Recall: 36.51% | Expected Conversions: 18.83
K=1500 | Precision: 1.46% | Recall: 42.84% | Expected Conversions: 21.92
K=2000 | Precision: 1.19% | Recall: 46.75% | Expected Conversions: 23.83

Test
K=500  | Precision: 3.93% | Recall: 34.15% | Expected Conversions: 19.67
K=1000 | Precision: 2.43% | Recall: 42.39% | Expected Conversions: 24.33
K=1500 | Precision: 1.82% | Recall: 47.66% | Expected Conversions: 27.33
K=2000 | Precision: 1.50% | Recall: 52.13% | Expected Conversions: 30.00
```

Качество этой модели чуть выше, чем у модели с заранее масштабированными признаками. Как и в случае с вариантом логистической регрессии, обученной на всех признаках, качество на train чуть лучше чем на test. Это ожидаемо из-за разного среднего кол-ва положительных классов по месяцам. Оценим веса модели.

In [10]:
print_lr_coefficients(lr_models_wo_scaled_featuers[0.1])

Топ-10 признаков с наибольшими coef_:
                         feature   coef
top_mcc_group_inn_Online&telecom 0.2395
              is_first_trx_month 0.1920
           top_mcc_group_inn_EDU 0.1768
         is_relevant_mcc_altpay5 0.1181
                    rule_flag_p3 0.1167
    is_relevant_turnover_altpay5 0.1128
                  turnover_cv_6m 0.0894
          became_active_recently 0.0794
cnt_trx_change_to_cnt_trx_ema_6m 0.0511
          has_recurrent_payments 0.0487

Топ-10 признаков с наименьшими coef_:
                        feature    coef
      top_mcc_group_inn_Charity -1.0884
     top_mcc_group_inn_FoodTech -0.8843
          has_termination_start -0.8786
               inn_status_Churn -0.4290
      top_mcc_group_inn_Grocery -0.4204
        top_mcc_group_inn_Churn -0.4190
who_is_this_first_inn_outcoming -0.1731
             is_relevant_mcc_p3 -0.1615
      lifetime_month_streak_inn -0.1281
         current_segment_inn_XS -0.1279

Доля признаков с коэффициентом 0: 52.70%



## LightGBM

In [17]:
from acquiring_xsell.preprocessing import train_test_month_split
from lightgbm import LGBMClassifier, early_stopping
from acquiring_xsell.metrics import calculate_metrics, precision_k


train, valid, test = train_test_month_split(df, get_valid=True)

X_train = train.drop(columns=['target', 'month'])
y_train = train['target']

X_valid = valid.drop(columns=['target', 'month'])
y_valid = valid['target']

X_test = test.drop(columns=['target', 'month'])
y_test = test['target']


lgbm = LGBMClassifier(objective="binary", random_state=RANDOM_STATE)
lgbm.fit(
    X_train, y_train,
    eval_set=[(X_valid, y_valid)],
    eval_metric="auc",  
    callbacks=[early_stopping(50)],
)

valid_tmp = valid.copy()
valid_tmp['lgbm_score'] = lgbm.predict_proba(X_valid)[:, 1]

test_tmp = test.copy()
test_tmp['lgbm_score'] = lgbm.predict_proba(X_test)[:, 1]

print("\nLightGBM P1P2")

print("\nValid")
calculate_metrics(
    valid_tmp,
    y_score="lgbm_score"
)

print("\nTest")
calculate_metrics(
    test_tmp,
    y_score="lgbm_score"
)

[LightGBM] [Info] Number of positive: 410, number of negative: 198901
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.014449 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 11629
[LightGBM] [Info] Number of data points in the train set: 199311, number of used features: 69
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.002057 -> initscore=-6.184405
[LightGBM] [Info] Start training from score -6.184405
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[2]	valid_0's auc: 0.752311	valid_0's binary_logloss: 0.0213272

LightGBM P1P2

Valid
K=500  | Precision: 4.67% | Recall: 36.19% | Expected Conversions: 23.33
K=1000 | Precision: 2.57% | Recall: 39.54% | Expected Conversions: 25.67
K=1500 | Precision: 1.93% | Recall: 44.99% | Expected Conversions: 29.00
K=2000 | Precision: 1.58% | Recall: 48.9

Попробуем подобрать гиперпараметры, чтобы улучшить качество.

In [11]:
from acquiring_xsell.preprocessing import train_test_month_split
from sklearn.model_selection import ParameterSampler
from lightgbm import LGBMClassifier, early_stopping
from acquiring_xsell.metrics import calculate_metrics, precision_k


train, valid, test = train_test_month_split(df, get_valid=True)

X_train = train.drop(columns=['target', 'month'])
y_train = train['target']

X_valid = valid.drop(columns=['target', 'month'])
y_valid = valid['target']

X_test = test.drop(columns=['target', 'month'])
y_test = test['target']


# pos_weight = (1 - 0.00206) / 0.00206 ≈ 486
# Перебираем вокруг этого значения + недо/перевес

param_grid = {
    "scale_pos_weight": [100, 250, 486, 700],
    "num_leaves": [31, 63, 127],
    "min_child_samples": [50, 100, 200],
    "subsample": [0.7, 0.9],
    "subsample_freq": [1],
    "colsample_bytree": [0.7, 0.9],
    "reg_lambda": [0.1, 1.0, 5.0],
    "learning_rate": [0.05],
    "n_estimators": [1000],
}

best_score = 0
best_params = None

for params in ParameterSampler(param_grid, n_iter=30, random_state=RANDOM_STATE):
    
    lgbm = LGBMClassifier(**params, objective="binary", random_state=RANDOM_STATE)
    lgbm.fit(
        X_train, y_train,
        eval_set=[(X_valid, y_valid)],
        eval_metric="auc",  
        callbacks=[early_stopping(50)],
    )

    valid_tmp = valid.copy()
    valid_tmp['lgbm_score'] = lgbm.predict_proba(X_valid)[:, 1]

    score = precision_k(valid_tmp, y_score_col='lgbm_score', k=2000)
    
    if score > best_score:
        best_score = score
        best_params = params
        best_lgbm = lgbm
    print(f"Current Best Precision@2000: {best_score:.2%}")


valid_tmp = valid.copy()
valid_tmp['lgbm_score'] = best_lgbm.predict_proba(X_valid)[:, 1]

test_tmp = test.copy()
test_tmp['lgbm_score'] = best_lgbm.predict_proba(X_test)[:, 1]

print("\nLightGBM P1P2")

print("\nValid")
calculate_metrics(
    valid_tmp,
    y_score="lgbm_score"
)

print("\nTest")
calculate_metrics(
    test_tmp,
    y_score="lgbm_score"
)

[LightGBM] [Info] Number of positive: 410, number of negative: 198901
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.015159 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 11629
[LightGBM] [Info] Number of data points in the train set: 199311, number of used features: 69
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.002057 -> initscore=-6.184405
[LightGBM] [Info] Start training from score -6.184405
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[1]	valid_0's auc: 0.724875	valid_0's binary_logloss: 4.98194
Current Best Precision@2000: 1.67%
[LightGBM] [Info] Number of positive: 410, number of negative: 198901
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.013090 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is

Пока качество у логистической регрессии выше, видимо в данных ярче выражены линейные зависимости.

## CatBoost

In [21]:
from acquiring_xsell.preprocessing import train_test_month_split
from sklearn.model_selection import ParameterSampler
from catboost import CatBoostClassifier
from acquiring_xsell.metrics import calculate_metrics, precision_k


train, valid, test = train_test_month_split(df, get_valid=True)

X_train = train.drop(columns=['target', 'month'])
y_train = train['target']

X_valid = valid.drop(columns=['target', 'month'])
y_valid = valid['target']

X_test = test.drop(columns=['target', 'month'])
y_test = test['target']

cat_cols = X_train.select_dtypes(include=['object', 'category']).columns.tolist()

cb = CatBoostClassifier(
    loss_function="Logloss",
    cat_features=cat_cols,
    random_seed=RANDOM_STATE,
    early_stopping_rounds=50,
    eval_metric="AUC",
    verbose=0,
)

cb.fit(
    X_train, y_train,
    eval_set=(X_valid, y_valid),
)

valid_tmp = valid.copy()
valid_tmp['cb_score'] = cb.predict_proba(X_valid)[:, 1]

valid_tmp['cb_score'] = cb.predict_proba(X_valid)[:, 1]
test_tmp['cb_score']  = cb.predict_proba(X_test)[:, 1]

print(f"CatBoost P1P2")

print("\nValid")
calculate_metrics(valid_tmp, y_score="cb_score")

print("\nTest")
calculate_metrics(test_tmp, y_score="cb_score")

CatBoost P1P2

Valid
K=500  | Precision: 4.80% | Recall: 35.99% | Expected Conversions: 24.00
K=1000 | Precision: 2.63% | Recall: 40.15% | Expected Conversions: 26.33
K=1500 | Precision: 2.04% | Recall: 47.59% | Expected Conversions: 30.67
K=2000 | Precision: 1.72% | Recall: 53.09% | Expected Conversions: 34.33

Test
K=500  | Precision: 3.13% | Recall: 27.24% | Expected Conversions: 15.67
K=1000 | Precision: 1.87% | Recall: 32.65% | Expected Conversions: 18.67
K=1500 | Precision: 1.53% | Recall: 40.33% | Expected Conversions: 23.00
K=2000 | Precision: 1.28% | Recall: 44.71% | Expected Conversions: 25.67


Попробуем подобрать гиперпараметры, чтобы улучшить качество.

In [12]:
from acquiring_xsell.preprocessing import train_test_month_split
from sklearn.model_selection import ParameterSampler
from catboost import CatBoostClassifier
from acquiring_xsell.metrics import calculate_metrics, precision_k


train, valid, test = train_test_month_split(df, get_valid=True)

X_train = train.drop(columns=['target', 'month'])
y_train = train['target']

X_valid = valid.drop(columns=['target', 'month'])
y_valid = valid['target']

X_test = test.drop(columns=['target', 'month'])
y_test = test['target']

cat_cols = X_train.select_dtypes(include=['object', 'category']).columns.tolist()

param_grid = {
    "scale_pos_weight": [100, 250, 486, 700],
    "depth": [4, 6, 8],
    "min_data_in_leaf": [50, 100, 200],
    "subsample": [0.7, 0.9],
    "colsample_bylevel":[0.7, 0.9],
    "l2_leaf_reg": [1, 3, 10, 30],
    "iterations": [1000],
    "learning_rate": [0.05],
}

best_score = 0
best_params = None
best_cb = None

for params in ParameterSampler(param_grid, n_iter=30, random_state=RANDOM_STATE):

    cb = CatBoostClassifier(
        **params,
        loss_function="Logloss",
        cat_features=cat_cols,
        random_seed=RANDOM_STATE,
        early_stopping_rounds=50,
        eval_metric="AUC",
        verbose=0,
    )

    cb.fit(
        X_train, y_train,
        eval_set=(X_valid, y_valid),
    )

    valid_tmp = valid.copy()
    valid_tmp['cb_score'] = cb.predict_proba(X_valid)[:, 1]

    score = precision_k(valid_tmp, y_score_col='cb_score', k=2000)

    if score > best_score:
        best_score = score
        best_params = params
        best_cb = cb

    print(f"Current Best Precision@2000: {best_score:.2%}")

print("\nBest params:", best_params)
print("Best iteration:", best_cb.get_best_iteration())

valid_tmp = valid.copy()
test_tmp = test.copy()

valid_tmp['cb_score'] = best_cb.predict_proba(X_valid)[:, 1]
test_tmp['cb_score']  = best_cb.predict_proba(X_test)[:, 1]

print(f"CatBoost P1P2")

print("\nValid")
calculate_metrics(valid_tmp, y_score="cb_score")

print("\nTest")
calculate_metrics(test_tmp, y_score="cb_score")

Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precision@2000: 1.83%
Current Best Precisi

Лучшее качество получилось на логистической регрессии без масштабированных признаков с L1-регуляризацией C=0.1

In [14]:
from joblib import dump
from pathlib import Path

best_model = lr_models_wo_scaled_featuers[0.1]
MODEL_PATH = Path("..", "models", "logreg_p1p2_wo_scaled_features.joblib")

dump(best_model, MODEL_PATH)

['../models/logreg_p1p2_wo_scaled_features.joblib']